src/main/java/org/spongepowered/common/mixin/core/entity/MixinEntity.java

```diff
@@ -360,6 +401,25 @@ public void postDealFireDamage(CallbackInfo callbackInfo) {
         }
     }
 
+    @Inject(method = "onStruckByLightning", at = @At(value = "FIELD", target = LIGHTNING_DAMAGESOURCE_FIELD, opcode = Opcodes.GETSTATIC))
+    public void preLightning(EntityLightningBolt lightningBolt, CallbackInfo callbackInfo) {
+        if (!this.world.isRemote) {
+            this.originalLightning = DamageSource.LIGHTNING_BOLT;
+            DamageSource.IN_FIRE = new EntityDamageSource("lightningBolt", lightningBolt);
+        }
+    }
+
+    @Inject(method = "onStruckByLightning", at = @At(value = "INVOKE_ASSIGN", target = ATTACK_ENTITY_FROM_METHOD))
+    public void postLightning(EntityLightningBolt lightningBolt, CallbackInfo callbackInfo) {
+        if (!this.world.isRemote) {
+            if (this.originalLightning == null) {
+                SpongeImpl.getLogger().error("Original lightning is null!");
+                Thread.dumpStack();
```

Comment:
This was mostly inspired by postDealFireDamage/postSetOnFire, which includes that for some reason.